stats/opentelemetry/client_metrics.go

```diff
@@ -0,0 +1,262 @@
+		return
+	}
+	csh.processRPCEvent(ctx, rs, ri.mi)
+}
+
+func (csh *clientStatsHandler) processRPCEvent(ctx context.Context, s stats.RPCStats, mi *metricsInfo) {
+	switch st := s.(type) {
+	case *stats.InHeader, *stats.OutHeader, *stats.InTrailer, *stats.OutTrailer:
+	case *stats.Begin:
+		ci := getCallInfo(ctx)
+		if ci == nil {
+			// Shouldn't happen, set by interceptor, defensive programming. Log it won't record?
+			return
+		}
+
+		if csh.registeredMetrics.clientAttemptStarted != nil {
+			csh.registeredMetrics.clientAttemptStarted.Add(ctx, 1, metric.WithAttributes(attribute.String("grpc.method", removeLeadingSlash(ci.method)), attribute.String("grpc.target", ci.target))) // Add records a change to the counter...attributeset for efficiency
+		}
+	case *stats.OutPayload:
+		atomic.AddInt64(&mi.sentCompressedBytes, int64(st.CompressedLength))
+	case *stats.InPayload:
+		atomic.AddInt64(&mi.recvCompressedBytes, int64(st.CompressedLength))
+	case *stats.End:
+		csh.processRPCEnd(ctx, mi, st)
+	default:
+		// Shouldn't happen. gRPC calls into stats handler, and will never not
+		// be one of the types above.
+		logger.Errorf("Received unexpected stats type (%T) with data: %v", s, s)
+	}
+}
+
+func (csh *clientStatsHandler) processRPCEnd(ctx context.Context, mi *metricsInfo, e *stats.End) {
+	ci := getCallInfo(ctx)
+	if ci == nil {
+		// Shouldn't happen, set by interceptor, defensive programming.
+		return
+	}
+	latency := float64(time.Since(mi.startTime)) / float64(time.Second)
+	var st string
+	if e.Error != nil {
+		s, _ := status.FromError(e.Error)
+		st = canonicalString(s.Code())
+	} else {
+		st = "OK"
+	}
+	clientAttributeOption := metric.WithAttributes(attribute.String("grpc.method", removeLeadingSlash(ci.method)), attribute.String("grpc.target", ci.target), attribute.String("grpc.status", st))
+	if csh.registeredMetrics.clientAttemptDuration != nil {
+		csh.registeredMetrics.clientAttemptDuration.Record(ctx, latency, clientAttributeOption)
+	}
+
+	if csh.registeredMetrics.clientAttemptSentTotalCompressedMessageSize != nil {
+		csh.registeredMetrics.clientAttemptSentTotalCompressedMessageSize.Record(ctx, atomic.LoadInt64(&mi.sentCompressedBytes), clientAttributeOption)
+	}
+
+	if csh.registeredMetrics.clientAttemptRcvdTotalCompressedMessageSize != nil {
+		csh.registeredMetrics.clientAttemptRcvdTotalCompressedMessageSize.Record(ctx, atomic.LoadInt64(&mi.recvCompressedBytes), clientAttributeOption)
+	}
+}
+
+// DefaultClientMetrics are the default client metrics provided by this module.
+var DefaultClientMetrics = Metrics{
+	metrics: map[string]bool{
+		"grpc.client.attempt.started":                            true,
+		"grpc.client.attempt.duration":                           true,
+		"grpc.client.attempt.sent_total_compressed_message_size": true,
+		"grpc.client.attempt.rcvd_total_compressed_message_size": true,
+		"grpc.client.call.duration":                              true,
+	},
```

Comment:
use `Metrics.Empty.Add()`? (Or create a `NewMetrics(...)`?)

Reply:
I would rather do the NewMetrics, but I think the decided API restricts the operations of construction enough (can't just pass in a list[], you have a default list you scale up and down by single metrics) that we should do a metrics.Empty.Add() here.

Reply:
`Add` has, and `NewMetrics` can have a variadic parameter.  So this could be `NewMetrics("grpc.client.attempt.started", "grpc.client.attempt.duration", ....)`.

Please use consts for these metric strings (as above).  I think we can just delete `EmptyMetrics`.

Reply:
I think NewMetrics having a variadic parameter breaks the API that we discussed...Add Remove or Empty() than add. I understand what you're going for, but in that case I might as well have passed a slice of metrics to initialize like I originally wanted but they pushed back wrt API on.

Reply:
I don't see how `EmptyMetrics.Add(...)` is any different than `NewMetrics(...)` except that one is much more obvious how to use it and doesn't have the weird loophole of someone doing (`otel.EmptyMetrics = <something else>`).

Reply:
Ping on this comment.

Reply:
Discussed offline; switched.